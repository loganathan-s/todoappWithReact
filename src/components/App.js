import React, { Component } from 'react';
import Header from './Header';
import Footer from './Footer';
import TaskForm from './TaskForm';
import Task from './Task';
import logo from '../images/logo.svg';
import {Tasks} from '../simpleTasks';

import '../css/App.css';

class App extends Component {
  constructor(){
    super();
    this.loadSamples = this.loadSamples.bind(this); 
    this.addTask = this.addTask.bind(this); 
    this.removeTask = this.removeTask.bind(this); 
    this.updateTask = this.updateTask.bind(this); 
    this.markComplete = this.markComplete.bind(this); 
    this.setCurrentTaskEditable = this.setCurrentTaskEditable.bind(this); 
    this.state = {
      tasks: {},
      editing: false,
      currentTask: {}
    };
  }

  addTask(task){
    const tasks = {...this.state.tasks};
    const taskKey = Object.keys(tasks).length + 1;
    tasks[`task${taskKey}`] = task;
    this.setState({ tasks })
  }

  removeTask(key){
     const tasks = {...this.state.tasks};
     delete tasks[key];
     this.setState({ tasks }) 
  }

  updateTask(key, updatedTask){
     const tasks = {...this.state.tasks};
     const task = tasks[key]
     task.task = updatedTask
     task.edit = false
     tasks[key] = task
     this.setState({ tasks }) 
  }

  markComplete(key){
     const tasks = {...this.state.tasks};
     let task = tasks[key]
     task.status = (task.status === "Done" ? "InProgress" : "Done")
     tasks[key] = task
     this.setState({ tasks }) 
  }

  setCurrentTaskEditable(key){
    const tasks = {...this.state.tasks};
    const currentTask = tasks[key]
    currentTask["edit"] = !currentTask["edit"];
    tasks[key] = currentTask;
    this.setState({ tasks });
  }

  loadSamples(){
      this.setState({
        tasks: Tasks
      })
  }

  //Callback which runs before rendering the APP component
  componentWillMount(){ 
    this.loadSamples();
  }

    
  render() {
    return (
      <div className="App">
         <Header />
          <div className="content">
            <TaskForm addTask={this.addTask} /> 
            <ul className="allTasks">
              { Object.
                keys(this.state.tasks).map(
                  key => <Task key={key} index={key} task={this.state.tasks[key]} removefromTask={this.removeTask} markasDone={this.markComplete}  setCurrentTaskEditable={this.setCurrentTaskEditable} updateTask={this.updateTask}/>)
            }
            </ul>
          </div>
        <Footer/>
      </div>
    );
  }
}

export default App;
