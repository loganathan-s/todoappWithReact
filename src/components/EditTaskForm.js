import React from 'react';
import PropTypes from 'prop-types';

class EditTaskForm extends React.Component{
	

	render(){
		const { task, index, setCurrentTaskEditable } = this.props
		return(
				<div>
				  <input ref={(input) => this.taskDesc = input } type="text" placeholder="Task"/>
				  <button type="submit" onClick={() => this.props.updateTask(index, this.taskDesc.value)}>Update</button>
				  <button onClick={() => setCurrentTaskEditable(index)}>Cancel</button>
				</div>
			)
	}
}

EditTaskForm.PropTypes = {
  task: PropTypes.object.isRequired,
  index: PropTypes.string.isRequired,
  setCurrentTaskEditable: PropTypes.func.isRequired,
  updateTask: PropTypes.func.isRequired,
};

export default EditTaskForm;