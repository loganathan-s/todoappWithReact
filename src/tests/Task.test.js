import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import App from '../components/App';
import Task from '../components/Task';
import EditTaskForm from '../components/EditTaskForm';


describe(Task, () => { 
	// Added to allow lifecycles, bydefault Enzyme wont execute lifecylce methods
	 const options = {
   	  	 lifecycleExperimental: true,
  		 disableLifecycleMethods: false 
	   };

	  const removeTask = jest.fn();	
	  const markComplete = jest.fn();
	  const setCurrentTaskEditable = jest.fn();
	  const updateTask = jest.fn();
	  const index = "task1";
	  const task =  {
        task: 'Everyones favorite white task. We will cut it to the size you need and ship it.',
        status: 'in_progress'
    };
    
	const appComponent = shallow(<App/>, options);
	const currentTaskKey = Object.keys(appComponent.state('tasks'))[0];
	let currentTask = appComponent.state('tasks')[Object.keys(appComponent.state('tasks'))[0]];
	const component = mount(<Task key={currentTaskKey} index={currentTaskKey} task={currentTask } removefromTask={removeTask} markasDone={markComplete}  setCurrentTaskEditable={setCurrentTaskEditable} updateTask={updateTask}/>);

	it('renders and matches the snapshot', () => {
    	const component = renderer.create(<Task key={index} index={index} task={ task } removefromTask={removeTask} markasDone={markComplete}  setCurrentTaskEditable={setCurrentTaskEditable} updateTask={updateTask}/>);
    	const tree = component.toJSON();
    	expect(tree).toMatchSnapshot();
    });


  	it('should not contain an EditTaskForm subcomponent', () => {
    	expect(component.find(EditTaskForm)).toHaveLength(0);
  	});

  	it('should contain Edit, MarkComplet and Remove Task Buttons', () => {
      expect(component.find('button')).toHaveLength(3);
  	});

  	it('should set the edit flag when I click edit button', () => {
  	    component.find('button.editTask').simulate('click');
        expect(setCurrentTaskEditable).toBeCalled();
        expect(currentTask.edit).toBe(undefined);
        appComponent.instance().setCurrentTaskEditable(index);
  	    expect(currentTask.edit).toBe(true);
  	});

    it('should set the task done flag when I click markComplete button', () => {
        expect(currentTask.status).toBe("in_progress");
        component.find('button.markcomplete').simulate('click');
        expect(markComplete).toBeCalled();
        appComponent.instance().markComplete(index);
        expect(currentTask.status).toBe("Done");
    });

    it('should set the task status to InProgress flag when I click markInProgress button', () => {
        component.find('button.markcomplete').simulate('click');
        expect(markComplete).toBeCalled();
        appComponent.instance().markComplete(index);
        expect(currentTask.status).toBe("InProgress");
    });

    it('should remove the task when removeTask button', () => {
        const totalTasks = Object.keys(appComponent.state('tasks')).length;
        component.find('button.removeTask').simulate('click');
        expect(removeTask).toBeCalled();
        appComponent.instance().removeTask(index);
        expect(Object.keys(appComponent.state('tasks')).length).toEqual(totalTasks - 1);
    });

    it('should display the Editform, when task has edit flag', () => {
        currentTask["edit"] = true;
        const component = mount(<Task key={currentTaskKey} index={currentTaskKey} task={currentTask } removefromTask={removeTask} markasDone={markComplete}  setCurrentTaskEditable={setCurrentTaskEditable} updateTask={updateTask}/>);
        expect(currentTask.edit).toBe(true);
        expect(component.find(EditTaskForm)).toHaveLength(1);
    });

});