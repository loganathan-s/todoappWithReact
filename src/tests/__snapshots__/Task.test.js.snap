// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Task extends _react2.default.Component {
	render() {
		const { task, index } = this.props; // Es6 Destructuring
		const markcompleteText = task.status === "Done" ? "Mark InProgress" : "Mark Complete";

		return (

			_react2.default.createElement('li', { className: 'task', __source: { fileName: _jsxFileName, lineNumber: 12 }, __self: this },
				_react2.default.createElement('h3', { className: 'taskName', __source: { fileName: _jsxFileName, lineNumber: 13 }, __self: this },
					task.task, ' - ', task.status),

				task.edit ? _react2.default.createElement(_EditTaskForm2.default, { task: task, index: index, setCurrentTaskEditable: this.props.setCurrentTaskEditable, updateTask: this.props.updateTask, __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this }) : _react2.default.createElement('div', { __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this }, _react2.default.createElement('button', { onClick: () => this.props.markasDone(index), className: 'markcomplete', __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this }, markcompleteText), _react2.default.createElement('button', { onClick: () => this.props.setCurrentTaskEditable(index), className: 'editTask', __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this }, 'Edit Task'), _react2.default.createElement('button', { onClick: () => this.props.removefromTask(index), className: 'removeTask', __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this }, 'RemoveTask'))));


	}} renders and matches the snapshot 1`] = `
<li
  className="task"
>
  <h3
    className="taskName"
  >
    Everyones favorite white task. We will cut it to the size you need and ship it.
     - 
    in_progress
  </h3>
  <div>
    <button
      className="markcomplete"
      onClick={[Function]}
    >
      Mark Complete
    </button>
    <button
      className="editTask"
      onClick={[Function]}
    >
      Edit Task
    </button>
    <button
      className="removeTask"
      onClick={[Function]}
    >
      RemoveTask
    </button>
  </div>
</li>
`;
